src/Columns/IColumn.cpp

```diff
@@ -83,6 +83,11 @@ ColumnPtr IColumn::createWithOffsets(const Offsets & offsets, const ColumnConst
     return res;
 }
 
+size_t IColumn::estimateCardinalityInPermutedRange(const IColumn::Permutation & /*permutation*/, const EqualRange & equal_range) const
```

Comment:
Having only implementations for String and Decimal columns is good to demonstrate concept but not enough for practical use cases. We'd return in almost all cases the same cardinalities which makes the row order optimization ineffective. We'd need implementations for ColumnString, ColumnFixedString, ColumnVector, ColumnNullable, ColumnLowCardinality. And possibly also for ColumnArray, ColumnTuple and ColumnMap (they are more advanced).

Reply:
I have implemented the **estimateCardinalityInPermutedRange** function for **ColumnDecimal**, **ColumnString**, **ColumnFixedString**, **ColumnVector**, **ColumnNullable**, **ColumnLowCardinality**.
For **ColumnArray**, **ColumnTuple** and **ColumnMap**, it seems to me, it is not worth to implement such logic (also, it may be necessary to remove it for **ColumnVector**).
The reason is that in reasonable tables, there should not be many identical arrays - in this case, split them into several columns and you will get speedup and compression improvement many times better than from my optimization. And, as it seems to me, we are writing code for real scenarios and rational users, not for an arbitrary case of random data in tables.

But, if you implement this function for arrays, you will find that they may end up in cardinality order before reasonable columns (numbers, string types), which will conversely worsen the result for “good” tables. Since lexicographic sorting of a large equivalence class of different rows, will give a better compression improvement than lexicographic sorting of a similar class of arrays of random objects.

Reply:
Okay, that is reasonable, thanks!